### PR TITLE
Update the crypto doc

### PR DESCRIPTION
**Pull request by tarakby**
Closes https://github.com/onflow/cadence/issues/1021

## Description

April mainnet spork updates the Flow Virtual Machine (FVM) implementation and some cryptography features in it. 
This PR updates the doc to match the new functionalities provided by Cadence and the FVM after the upcoming spork. 
It also provides more details on how to generate signatures that Cadence/FVM consider valid.

### CODE REVIEW

File: docs/language/crypto.mdx [hunk 1]
```diff
@@ -169,57 +219,98 @@ let isValid = pk.verify(
...
+### BLS signature aggregation
+
+`cadence•fun aggregateSignatures(_ signatures: [[UInt8]]): [UInt8]?`
+
+Aggregates multiple BLS signatures into one. 
+Signatures could be generated from the same or distinct messages, they
+could also be the aggregation of other signatures.
+The order of the signatures in the slice does not matter since the aggregation is commutative.
+There is no subgroup membership check performed on the input signatures.
+The function errors if the array is empty or if decoding one of the signature fails.
```
Comment (tarakby) [inline docs/language/crypto.mdx:294]: Not sure whether "errors" is the correct word.

File: docs/language/crypto.mdx [hunk 2]
```diff
@@ -169,57 +219,98 @@ let isValid = pk.verify(
...
+`cadence•fun aggregatePublicKeys(_ publicKeys: [PublicKey]): PublicKey?`
 
-- `cadence•fun aggregatePublicKeys(_ publicKeys: [PublicKey]): PublicKey?`
+Aggregates multiple BLS public keys into one.
 
-  Aggregates multiple BLS public keys into one.
+The order of the public keys in the slice does not matter since the aggregation is commutative.
+The input keys are guaranteed to be in the correct subgroup since subgroup membership is checked
+at the key creatiom time. 
+The function errors if the array is empty or any of the input keys is not a BLS key.
```
Comment (tarakby) [inline docs/language/crypto.mdx:309]: Not sure whether "errors" is the correct word.

File: docs/language/crypto.mdx [hunk 3]
```diff
@@ -107,11 +127,23 @@ struct PublicKey {
     ): Bool
 
     /// Verifies the proof of possession of the private key.
-    /// This function is only implemented if the signature algorithm of the public key is BLS, it errors if called with any other signature algorithm. 
+    /// This function is only implemented if the signature algorithm of the public key is BLS (BLS_BLS12_381).
+    /// It errors if called with any other signature algorithm.
```
Comment (turbolent) [inline docs/language/crypto.mdx:131]:
```suggestion
    /// If called with any other signature algorithm, the program aborts 
```

File: docs/language/crypto.mdx (hunk 1 above)
Comment (turbolent) [inline docs/language/crypto.mdx:294]:
```suggestion
If the array is empty or if decoding one of the signature fails, the program aborts
```

File: docs/language/crypto.mdx (hunk 2 above)
Comment (turbolent) [inline docs/language/crypto.mdx:309]:
```suggestion
If the array is empty or any of the input keys is not a BLS key, the program aborts
```

File: docs/language/crypto.mdx (hunk 3 above)
Comment (turbolent) [reply docs/language/crypto.mdx:131]: Also, please update the docstring in the code, here: https://github.com/onflow/cadence/blob/master/runtime/sema/type.go#L6260

File: docs/language/crypto.mdx [hunk 4]
```diff
@@ -107,11 +127,23 @@ struct PublicKey {
...
+    /// It errors if called with any other signature algorithm. 
     pub fun verifyPoP(_ proof: [UInt8]): Bool
 }
 ```
 
+<Callout type="info">
+
+🚧 Status: VerifyPoP is going to be available in an upcoming release of Cadence
+
+</Callout>
```
Comment (turbolent) [inline docs/language/crypto.mdx:140]: This should be released now, correct?

```suggestion
```

File: docs/language/crypto.mdx [hunk 5]
```diff
@@ -135,8 +167,26 @@ The raw key value depends on the supported signature scheme:
   [IETF draft-irtf-cfrg-pairing-friendly-curves-08](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-08.html#name-point-serialization-procedu).
   A public key is 96-bytes long.
 
+### Public Key validation
+
+<Callout type="info">
+
+🚧 Status: Validation at creation time is going to be available in an upcoming release of Cadence
```
Comment (turbolent) [inline docs/language/crypto.mdx:174]:
```suggestion
```

File: docs/language/crypto.mdx [hunk 6]
```diff
@@ -135,8 +167,26 @@ The raw key value depends on the supported signature scheme:
...
 
-A public key is validated at the time of creation. Hence, public keys are immutable.
+</Callout>
```
Comment (turbolent) [inline docs/language/crypto.mdx:176]:
```suggestion
```

File: docs/language/crypto.mdx [hunk 7]
```diff
@@ -169,57 +219,98 @@ let isValid = pk.verify(
...
+  The signature is 64 bytes-long for both curves. 
+  - `signedData` is the arbitrary message to verify the signature against.
+  - `domainSeparationTag` is the expected domain tag. Multiple valid tag values can be used (check [`hashWithTag`](#hashing) for more details).
+  - `hashAlgorithm` is either `SHA2_256`, `SHA3_256` or `KECCAK_256`. It is the algorithm used to hash the message along with the given tag (check the `hashWithTag`[function](#hashing) for more details).
+
+<Callout type="info">
+
+🚧 Status: ECDSA verificaction with any tag is going to be available in an upcoming release of Cadence
+
+</Callout>
```
Comment (turbolent) [inline docs/language/crypto.mdx:236]: This should be released now

```suggestion
```

File: docs/language/crypto.mdx [hunk 8]
```diff
@@ -169,57 +219,98 @@ let isValid = pk.verify(
...
+A valid signature would be generated using the expected `signedData` and `domainSeparationTag`, as well the same hashing to curve process. 
 
-## BLS
+## BLS multi-signature
 
 <Callout type="info">
 
 🚧 Status: This API is going to be available in an upcoming release of Cadence
 
 </Callout>
```
Comment (turbolent) [inline docs/language/crypto.mdx:266]: This should be released now, right?

```suggestion
```

**Review by turbolent — COMMENTED** [review]
Looks good, just some small suggestions to keep wording consistent across the API

File: docs/language/crypto.mdx (hunk 4 above)
Comment (tarakby) [reply docs/language/crypto.mdx:140]: Yes, I thought this PR will get merged earlier, and then I can remove these callouts after the spork :) 
I'll remove them now.

**Comment by tarakby** [comment]
@turbolent I think I've addressed your comments, but please let me know if more changes are needed.

**Review by turbolent — APPROVED** [review]
Looks good, thank you for improving this! 👍

**Pull request merged**